core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/DisposalSupport.java

```diff
@@ -0,0 +1,31 @@
+package org.optaplanner.core.impl.solver;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The class is not thread-safe and is meant to be accessed from the solver thread only.
+ */
+public final class DisposalSupport {
+
+    private final List<AutoCloseable> disposables = new ArrayList<>();
+
+    public void addDisposable(AutoCloseable disposable) {
+        disposables.add(disposable);
+    }
+
+    public void removeDisposable(AutoCloseable disposable) {
+        disposables.remove(disposable);
+    }
+
+    public void dispose() {
+        for (AutoCloseable disposable : disposables) {
+            try {
+                disposable.close();
+            } catch (Exception e) {
+                throw new IllegalStateException("Failed to dispose a resource (" + disposable + ").", e);
```

Comment:
Not sure about this. Do we really want to kill the entire application because something failed to close?

Reply:
As an alternative, the Solver could emit an event that an exception occurred, leaving the decision whether to dispose of something or not to a listener.

Reply:
IMO there is nothing wrong with the behavior we had until now - if the phase fails, the exception is thrown. We just want to prevent the JVM from hanging. I really do not see the need to drag any of this into the solver.

Reply:
Agreed on the alternative to add `solvingError` to the `SolverLifecycleListener` and let the individual components properly react on this event.